### PR TITLE
[EDL] Improve Documentation & Add additional HTTP Header

### PR DESCRIPTION
**Pull request by MichaelYochpaz**
## Status
- [ ] In Progress
- [x] Ready
- [ ] In Hold - (Reason for hold)

## Related Issues
relates: [link to the issue](https://jira-hq.paloaltonetworks.local/browse/XSUP-23560)

## Description
Add a new `X-EDL-Origin-Size` HTTP header to responses, and improve the documentation.

### CODE REVIEW

File: Packs/EDL/Integrations/EDL/EDL.py [hunk 1]
```diff
@@ -857,49 +880,62 @@ def route_edl() -> Response:
...
     if username and password:
         headers: dict = cast(Dict[Any, Any], request.headers)
         if not validate_basic_authentication(headers, username, password):
             err_msg: str = 'Basic authentication failed. Make sure you are using the right credentials.'
             demisto.debug(err_msg)
-            return Response(err_msg, status=401, mimetype='text/plain', headers=[
-                ('WWW-Authenticate', 'Basic realm="Login Required"'),
-            ])
+            return Response(err_msg, status=401, mimetype='text/plain',
+                            headers=[('WWW-Authenticate', 'Basic realm="Login Required"')])
```
Comment (DeanArbel) [inline Packs/EDL/Integrations/EDL/EDL.py:890]: please revert

File: Packs/EDL/Integrations/EDL/EDL.py [hunk 2]
```diff
@@ -785,35 +789,54 @@ def get_outbound_mimetype(request_args: RequestArguments) -> str:
...
+            and the number of original indicators received from the server before formatting (int).
     """
+    global EDL_ON_DEMAND_CACHE_ORIGINAL_SIZE
     ctx = get_integration_context()
+
     if EDL_ON_DEMAND_KEY in ctx:
         ctx.pop(EDL_ON_DEMAND_KEY, None)
         request_args = RequestArguments.from_context_json(ctx)
-        edl = create_new_edl(request_args)
+        edl_data, original_indicators_count = create_new_edl(request_args)
```
Comment (DeanArbel) [inline Packs/EDL/Integrations/EDL/EDL.py:807]: let's remove `original_indicators_count` to simplify the code flow, and use just `EDL_ON_DEMAND_CACHE_ORIGINAL_SIZE` - even if `file.write(edl_data)` fails, it'll have to recreate the EDL, so it's the same effect.

File: Packs/EDL/Integrations/EDL/EDL.py [hunk 3]
```diff
@@ -857,49 +880,62 @@ def route_edl() -> Response:
...
+                            headers=[('WWW-Authenticate', 'Basic realm="Login Required"')])
 
     request_args = get_request_args(request.args, params)
     on_demand = params.get('on_demand')
     created = datetime.now(timezone.utc)
-    edl = get_edl_on_demand() if on_demand else create_new_edl(request_args)
-    etag = f'"{hashlib.sha1(edl.encode()).hexdigest()}"'    # nosec
+    edl_data, original_indicators_count = get_edl_on_demand() if on_demand else create_new_edl(request_args)
     query_time = (datetime.now(timezone.utc) - created).total_seconds()
+    etag = f'"{hashlib.sha1(edl_data.encode()).hexdigest()}"'  # nosec
```
Comment (DeanArbel) [inline Packs/EDL/Integrations/EDL/EDL.py:897]: please revert

**Review by DeanArbel — CHANGES_REQUESTED** [review]
Looks good, just a couple of notes.

File: Packs/EDL/Integrations/EDL/EDL.py (hunk 3 above)
Comment (MichaelYochpaz) [reply Packs/EDL/Integrations/EDL/EDL.py:897]: I didn't actually add this line as GitHub suggests, I just moved it down a line.

The code is:
```python
    created = datetime.now(timezone.utc)
    edl_data, original_indicators_count = get_edl_on_demand() if on_demand else create_new_edl(request_args)
    query_time = (datetime.now(timezone.utc) - created).total_seconds()
    etag = f'"{hashlib.sha1(edl_data.encode()).hexdigest()}"'  # nosec
    ```
    
Seems to make more sense to me that `query_time` is called right after calling the query, as it's a runtime counter.

File: Packs/EDL/Integrations/EDL/EDL.py (hunk 1 above)
Comment (MichaelYochpaz) [reply Packs/EDL/Integrations/EDL/EDL.py:890]:
```suggestion
            return Response(err_msg, status=401, mimetype='text/plain', headers=[
                ('WWW-Authenticate', 'Basic realm="Login Required"'),
            ])
```

File: Packs/EDL/ReleaseNotes/3_1_24.md [hunk 4]
```diff
@@ -0,0 +1,8 @@
+
+#### Integrations
+
+##### Generic Export Indicators Service
+
+- Added a new `X-EDL-Start-Size` header for EDL HTTP responses that can be used for debugging.
```
Comment (DeanArbel) [inline Packs/EDL/ReleaseNotes/3_1_24.md:6]: Please explain what it returns

File: Packs/EDL/Integrations/EDL/EDL.py [hunk 5]
```diff
@@ -869,37 +889,50 @@ def route_edl() -> Response:
...
     max_age = ceil((datetime.now() - dateparser.parse(cache_refresh_rate)).total_seconds())  # type: ignore[operator]
-    demisto.debug(f'edl: Returning edl of size: [{edl_size}], created: [{created}], query time seconds: [{query_time}],'
-                  f' max age: [{max_age}], etag: [{etag}]')
-    resp = Response(edl, status=200, mimetype=mimetype, headers=[
+
+    headers = [
         ('X-EDL-Created', created.isoformat()),
         ('X-EDL-Query-Time-Secs', "{:.3f}".format(query_time)),
         ('X-EDL-Size', str(edl_size)),
+        ('X-EDL-Start-Size', original_indicators_count),  # TODO
```
[the remark this answers is not in the document]
Comment (MichaelYochpaz) [reply Packs/EDL/Integrations/EDL/EDL.py:924]: Oops :)

File: Packs/EDL/ReleaseNotes/3_1_24.md (hunk 4 above)
Comment (MichaelYochpaz) [reply Packs/EDL/ReleaseNotes/3_1_24.md:6]:
```suggestion
- Added a new `X-EDL-Start-Size` header for HTTP responses representing the amount of indicators fetched before formatting. (The amount of indicators post-formatting is represented by the existing `X-EDL-Size` header.)
```

**Review by DeanArbel — APPROVED** [review]
Nicely done

File: Packs/EDL/Integrations/EDL/README.md [hunk 6]
```diff
@@ -200,13 +200,27 @@ There is no context output for this command.
...
 'EDL will be updated the next time you access it'
 
-### Troubleshooting
-Memory issue can happen in CSV / JSON format over 150,000 if all fields are selected.
 
-#### In terms of times
-* 10,000 indicators can take 10 - 20 seconds.
-* 100,000 indicators can take 1 - 3 minutes.
+### Troubleshooting
+* Indicators that are passed through the integration undergo formatting and deduplication process, which may lead to an apparent loss of indicators.
```
Comment (ShirleyDenkberg) [inline Packs/EDL/Integrations/EDL/README.md:205]:
```suggestion
* Indicators that are passed through the integration undergo formatting and deduplication, which may lead to an apparent loss of indicators.  
```

**Comment by ShirleyDenkberg** [comment]
@DeanArbel @glicht @bziser @MLainer1   Doc review completed.

**Review by bziser — APPROVED** [review]
great

**Pull request merged**